// template-picker/picker.js
var fs = require('fs'),
	path = require('path'),
	nf = require('node-file')
	;
var jstemplate = require('./jstemplate.js'),
	ztool = require('./ztool.js')
	;

var DEFAULT_CONFIG = {
	source: null,
	target: './',
	templateOutput: null,
	templatePlaceholder: '/*%HtmlTemplateFunctions%*/',
	compressWhitespace: false
}

var tmplRegexp = /<script\s*id="([^"]+)"\s*type="text\/template"\s*>([\s\S]*?)<\/script>/ig;
var tmplRegexp2 = /<script\s*type="text\/template"\s*id="([^"]+)"\s*>([\s\S]*?)<\/script>/ig;

var result;

var readConfig = function(config){
	if(ztool.isString(config)){
        var content = fs.readFileSync(config).toString();
        config = ztool.jsonParse(content);
    }
    config = ztool.merge({}, DEFAULT_CONFIG, config);
    if(ztool.isString(config.source)){
    	config.source = [config.source];
    }
    var source = [], list;
    for(var i = 0, s; s = config.source[i]; i++){
    	if(nf.isDirectoryPath(s)){
    		list = nf.listFilesSync(s, 'html, htm');
    		for(var j = 0, l; l = list[j]; j++){
    			source.push(path.join(s, l));
    		}
    	}else{
    		source.push(s);
    	}
    }
    config.source = source;
    // console.log(source);
    return config;
}

var pickup = function(match, tid, tmpl){
	tmpl = tmpl.replace(/\n|\r/g, '');
	var func = jstemplate.compile(tmpl);
	func = func.toString();
	var prefix = 'function anonymous';
	if(func.indexOf(prefix) == 0) {
		func = func.replace(prefix, 'function');
	}
	func = 'this.' + tid + '=' + func;
	result.push(func);
	return '';
}

var exec = function(configFile){
	var content, fileName;
	var config = readConfig(configFile);
	result = [];
	// console.log(config);
	for(var i = 0; i < config.source.length; i++) {
		fileName = config.source[i];
		content = fs.readFileSync(fileName).toString();
		content = content.replace(tmplRegexp, pickup);
		content = content.replace(tmplRegexp2, pickup);

		fileName = path.join(config.target, path.basename(fileName));
		console.log(fileName);
		nf.writeFileSync(fileName, content);
	}
	content = result.join('\n');
	if(config.compressWhitespace){
		content = content.replace(/[\t ]+/g, ' ');
	}
	var tmpl = fs.readFileSync(config.templateOutput).toString();
	tmpl = tmpl.replace(config.templatePlaceholder, content);
	fileName = path.join(config.target, path.basename(config.templateOutput));
	nf.writeFileSync(fileName, tmpl);
}

if(process.argv.length > 2 && 
   process.argv[1].toString() === 'template-picker'){
	exec(process.argv[2]);
}else{
	//非命令行启动，而是 js 的require
	exports.exec = exec;
}
